File: crypto-bot.js
const TelegramBot = require('node-telegram-bot-api');
const axios = require('axios');
const fs = require('fs');
const path = require('path');
require('dotenv').config();

// PID file for instance management
const PID_FILE = path.join(__dirname, 'bot.pid');

// Store bot instance globally
let botInstance = null;

// Bot configuration
const TOKEN = process.env.TELEGRAM_BOT_TOKEN || '8014754930:AAG3FxXaw4AAiJfP5dPmoV5HoqZirR6Qvwc';
const PORT = process.env.PORT || 3000;
const CHANNEL = process.env.TELEGRAM_CHANNEL || '@cryptoprices254';
const PROMOTED_CHANNEL = process.env.PROMOTED_CHANNEL || '@legitairdropsfb';
const CMC_API_KEY = process.env.CMC_API_KEY || 'b54bcf4d-1bca-4e8e-9a24-22ff2c3d462c'; // Demo key

// Auto-posting interval (5 minutes to avoid rate limits)
let channelPostingInterval = null;

// CoinMarketCap API configuration
const API_URL = "https://pro-api.coinmarketcap.com/v1/cryptocurrency/listings/latest";
const API_PARAMS = {
    start: 1,
    limit: 10,
    convert: 'USD'
};

// Check for existing instance
function checkExistingInstance() {
    if (fs.existsSync(PID_FILE)) {
        try {
            const existingPid = fs.readFileSync(PID_FILE, 'utf8').trim();
            // Check if process is still running
            process.kill(existingPid, 0);
            log(`❌ Another bot instance is already running (PID: ${existingPid})`);
            log('🛑 Please stop the existing instance first or wait for it to finish');
            process.exit(1);
        } catch (e) {
            // Process doesn't exist, remove stale PID file
            fs.unlinkSync(PID_FILE);
            log('🗑️ Removed stale PID file');
        }
    }
}

// Create PID file
function createPidFile() {
    fs.writeFileSync(PID_FILE, process.pid.toString());
    log(`📝 Created PID file: ${process.pid}`);
}

// Remove PID file
function removePidFile() {
    if (fs.existsSync(PID_FILE)) {
        fs.unlinkSync(PID_FILE);
        log('🗑️ Removed PID file');
    }
}

// Create bot instance with error handling
function createBot() {
    try {
        if (botInstance) {
            log('🛑 Stopping existing bot instance...');
            botInstance.stopPolling();
            botInstance = null;
        }
        
        botInstance = new TelegramBot(TOKEN, { 
            polling: {
                interval: 1000,
                autoStart: false,
                params: {
                    timeout: 10
                }
            }
        });
        
        // Add error handlers immediately
        botInstance.on('error', (error) => {
            log(`❌ Bot error: ${error.message}`);
        });

        botInstance.on('polling_error', (error) => {
            log(`❌ Polling error: ${error.message}`);
            if (error.message.includes('409') || error.message.includes('Conflict')) {
                log('🔄 Attempting to restart polling...');
                setTimeout(() => {
                    if (botInstance) {
                        botInstance.stopPolling();
                        setTimeout(() => {
                            if (botInstance) {
                                botInstance.startPolling();
                            }
                        }, 2000);
                    }
                }, 1000);
            }
        });
        
        return botInstance;
    } catch (error) {
        log(`❌ Failed to create bot: ${error.message}`);
        throw error;
    }
}

// Store active jobs
const activeJobs = new Map();

// Logging function
function log(message) {
    console.log(`[${new Date().toISOString()}] ${message}`);
}

// Fetch cryptocurrency data from CoinMarketCap API
async function getCryptoData() {
    try {
        log('🔄 Fetching crypto data from CoinMarketCap...');
        
        const response = await axios.get(API_URL, {
            params: API_PARAMS,
            timeout: 15000,
            headers: {
                'X-CMC_PRO_API_KEY': CMC_API_KEY,
                'Accept': 'application/json'
            }
        });
        
        log(`📡 API Response: ${response.status}`);
        
        if (response.status === 200) {
            const data = response.data.data;
            // Filter out USDT and get top 5
            const filteredData = data.filter(crypto => 
                crypto.symbol && crypto.symbol !== 'USDT'
            );
            const result = filteredData.slice(0, 5);
            log(`✅ Successfully fetched ${result.length} cryptocurrencies`);
            return result;
        } else {
            log(`❌ API Error: ${response.status}`);
            return [];
        }
    } catch (error) {
        log(`❌ Exception in getCryptoData: ${error.message}`);
        return [];
    }
}

// Format cryptocurrency data into a message
function formatCryptoMessage(cryptos) {
    if (!cryptos || cryptos.length === 0) {
        return "❌ Unable to fetch cryptocurrency data. Please try again later.";
    }
    
    let message = "🚀 **Top 5 Cryptocurrencies** 🚀\n\n";
    
    cryptos.forEach((crypto, index) => {
        const name = crypto.name || 'Unknown';
        const symbol = crypto.symbol || '';
        const price = crypto.quote?.USD?.price || 0;
        const change24h = crypto.quote?.USD?.percent_change_24h || 0;
        
        // Format price
        let priceStr;
        if (price >= 1) {
            priceStr = `$${price.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
        } else {
            priceStr = `$${price.toFixed(6)}`;
        }
        
        // Format change with emoji
        let changeStr;
        if (change24h > 0) {
            changeStr = `📈 +${change24h.toFixed(2)}%`;
        } else if (change24h < 0) {
            changeStr = `📉 ${change24h.toFixed(2)}%`;
        } else {
            changeStr = `➡️ ${change24h.toFixed(2)}%`;
        }
        
        message += `${index + 1}. **${name}** (${symbol})\n`;
        message += `   💰 ${priceStr}\n`;
        message += `   📊 ${changeStr}\n\n`;
    });
    
    const now = new Date();
    message += `🔄 Updated at ${now.toLocaleTimeString()}`;
    message += `\n\n💎 Join ${PROMOTED_CHANNEL} for exclusive airdrops and crypto signals!`;
    message += `\n🚀 Follow us for daily crypto updates and opportunities!`;
    
    return message;
}

// Send crypto update to channel
async function sendCryptoUpdate(chatId) {
    try {
        log('📊 Sending crypto update...');
        
        const cryptos = await getCryptoData();
        const message = formatCryptoMessage(cryptos);
        
        await botInstance.sendMessage(chatId, message, { 
            parse_mode: 'Markdown',
            disable_web_page_preview: true
        });
        
        log('✅ Crypto update sent successfully!');
        
    } catch (error) {
        log(`❌ Error sending crypto update: ${error.message}`);
        try {
            await botInstance.sendMessage(chatId, "❌ Error updating crypto prices. Will retry soon.");
        } catch (e) {
            log(`❌ Failed to send error message: ${e.message}`);
        }
    }
}

// Auto-post to channel function
async function autoPostToChannel() {
    try {
        log(`📢 Auto-posting to ${CHANNEL}...`);
        
        const cryptos = await getCryptoData();
        const message = formatCryptoMessage(cryptos);
        
        // Post to the channel
        await botInstance.sendMessage(CHANNEL, message, { 
            parse_mode: 'Markdown',
            disable_web_page_preview: true
        });
        
        log(`✅ Auto-posted to ${CHANNEL} successfully`);
        
    } catch (error) {
        log(`❌ Error auto-posting to channel: ${error.message}`);
    }
}

// Start auto-posting to channel
function startChannelPosting() {
    if (channelPostingInterval) {
        clearInterval(channelPostingInterval);
    }
    
    // Post immediately, then every 2 minutes
    autoPostToChannel();
    channelPostingInterval = setInterval(autoPostToChannel, 120000); // 2 minutes
    log(`📢 Auto-posting to ${CHANNEL} every 2 minutes...`);
}

// Stop auto-posting to channel
function stopChannelPosting() {
    if (channelPostingInterval) {
        clearInterval(channelPostingInterval);
        channelPostingInterval = null;
        log('🛑 Stopped auto-posting to channel');
    }
}
// Start command handler
function setupHandlers() {
    if (!botInstance) return;
    
    botInstance.onText(/\/start/, async (msg) => {
    try {
        const chatId = msg.chat.id;
        const userName = msg.from.first_name || "User";
        
        log(`🚀 /start command from ${userName} (chat_id: ${chatId})`);
        
        // Clear existing interval if any
        if (activeJobs.has(chatId)) {
            clearInterval(activeJobs.get(chatId));
            log('🗑️ Removed existing job');
        }
        
        // Schedule new job - every 60 seconds
        const intervalId = setInterval(() => {
            sendCryptoUpdate(chatId);
        }, 120000); // 2 minutes
        
        activeJobs.set(chatId, intervalId);
        
        const welcomeMsg = `👋 Hello ${userName}!\n\n` +
            `🤖 **Crypto Price Bot** is now active!\n\n` +
            `📊 You'll receive top 5 crypto prices every 2 minutes\n` +
            `🔄 First update in 3 seconds...\n\n` +
            `**Commands:**\n` +
            `• /start - Start updates\n` +
            `• /stop - Stop updates\n` +
            `• /prices - Get prices now\n` +
            `• /channel - Post to ${CHANNEL}\n\n` +
            `💎 Join ${PROMOTED_CHANNEL} for exclusive airdrops and crypto signals!`;
        
        await botInstance.sendMessage(chatId, welcomeMsg, { parse_mode: 'Markdown' });
        
        // Send first update after 3 seconds
        setTimeout(() => {
            sendCryptoUpdate(chatId);
        }, 3000);
        
        log('✅ Welcome message sent');
        
    } catch (error) {
        log(`❌ Error in start command: ${error.message}`);
        botInstance.sendMessage(msg.chat.id, "❌ Error starting bot. Please try again.");
    }
    });

    // Stop command handler
    botInstance.onText(/\/stop/, async (msg) => {
    try {
        const chatId = msg.chat.id;
        log(`🛑 /stop command from chat_id: ${chatId}`);
        
        if (activeJobs.has(chatId)) {
            clearInterval(activeJobs.get(chatId));
            activeJobs.delete(chatId);
            await botInstance.sendMessage(chatId, "🛑 **Updates stopped!** Use /start to resume.");
            log('✅ Jobs stopped successfully');
        } else {
            await botInstance.sendMessage(chatId, "ℹ️ No active updates to stop.");
        }
        
    } catch (error) {
        log(`❌ Error in stop command: ${error.message}`);
        botInstance.sendMessage(msg.chat.id, "❌ Error stopping updates.");
    }
    });

    // Prices command handler - immediate update
    botInstance.onText(/\/prices/, async (msg) => {
    try {
        const chatId = msg.chat.id;
        log(`💰 /prices command from chat_id: ${chatId}`);
        
        await botInstance.sendMessage(chatId, "🔄 Fetching current prices...");
        await sendCryptoUpdate(chatId);
        
        log('✅ Immediate prices sent');
        
    } catch (error) {
        log(`❌ Error in prices command: ${error.message}`);
        botInstance.sendMessage(msg.chat.id, "❌ Error fetching prices.");
    }
    });

    // Channel command handler - post to @legitairdropsfb
    botInstance.onText(/\/channel/, async (msg) => {
    try {
        const chatId = msg.chat.id;
        log(`📢 /channel command from chat_id: ${chatId}`);
        
        await botInstance.sendMessage(chatId, `🔄 Posting to ${CHANNEL}...`);
        
        const cryptos = await getCryptoData();
        const message = formatCryptoMessage(cryptos);
        
        // Post to the channel
        await botInstance.sendMessage(CHANNEL, message, { 
            parse_mode: 'Markdown',
            disable_web_page_preview: true
        });
        
        await botInstance.sendMessage(chatId, `✅ Successfully posted to ${CHANNEL}!`);
        log(`✅ Posted to ${CHANNEL} successfully`);
        
    } catch (error) {
        log(`❌ Error posting to channel: ${error.message}`);
        botInstance.sendMessage(msg.chat.id, `❌ Error posting to ${CHANNEL}. Make sure the bot is an admin in the channel.`);
    }
    });
}

// Test API and start bot
async function main() {
    try {
        // Check for existing instance
        checkExistingInstance();
        
        // Create PID file
        createPidFile();
        
        log('🚀 Starting Telegram Crypto Bot...');
        
        // Create bot instance
        createBot();
        
        // Setup command handlers
        setupHandlers();
        
        // Test API first
        log('🧪 Testing CoinMarketCap API...');
        const testData = await getCryptoData();
        if (testData && testData.length > 0) {
            log(`✅ API test successful - got ${testData.length} cryptos`);
        } else {
            log('⚠️ API test failed, but continuing...');
        }
        
        // Test bot token
        const me = await botInstance.getMe();
        log(`✅ Bot connected: @${me.username}`);
        
        // Start auto-posting to channel
        startChannelPosting();
        
        // Start polling
        await botInstance.startPolling();
        
        // Keep-alive server for Railway
        const http = require('http');
        const server = http.createServer((req, res) => {
            res.writeHead(200, { 'Content-Type': 'text/plain' });
            res.end('Crypto Bot is running! 🚀\n');
        });
        server.listen(PORT, () => {
            log(`🌐 Keep-alive server running on port ${PORT}`);
        });
        
        log(`📢 Target channel: ${CHANNEL}`);
        log(`🎯 Promoted channel: ${PROMOTED_CHANNEL}`);
        log('🔄 Bot is now running and listening for commands...');
        log(`🌐 Server ready on port ${PORT}`);
        log('');
        log('Available commands:');
        log('• /start - Start receiving crypto updates');
        log('• /stop - Stop updates');
        log('• /prices - Get current prices');
        log(`• /channel - Post update to ${CHANNEL}`);
        log('');
        log(`📢 Auto-posting to ${CHANNEL} every 2 minutes promoting ${PROMOTED_CHANNEL}`);
        
    } catch (error) {
        log(`❌ Failed to start bot: ${error.message}`);
        removePidFile();
        process.exit(1);
    }
}

// Graceful shutdown
function gracefulShutdown(signal) {
    log('🛑 Shutting down bot...');
    
    // Stop channel posting
    stopChannelPosting();
    
    // Clear all active jobs
    activeJobs.forEach((intervalId) => {
        clearInterval(intervalId);
    });
    activeJobs.clear();
    
    // Stop bot polling
    if (botInstance) {
        botInstance.stopPolling();
        botInstance = null;
    }
    
    // Remove PID file
    removePidFile();
    
    log(`✅ Bot shutdown complete (${signal})`);
    process.exit(0);
}

// Handle various shutdown signals
process.on('SIGINT', () => gracefulShutdown('SIGINT'));
process.on('SIGTERM', () => gracefulShutdown('SIGTERM'));
process.on('SIGUSR2', () => gracefulShutdown('SIGUSR2')); // nodemon restart

// Handle uncaught exceptions
process.on('uncaughtException', (error) => {
    log(`❌ Uncaught Exception: ${error.message}`);
    removePidFile();
    process.exit(1);
});

process.on('unhandledRejection', (reason, promise) => {
    log(`❌ Unhandled Rejection at: ${promise}, reason: ${reason}`);
    removePidFile();
    process.exit(1);
});

// Start the bot
main();

File: crypto-bot.py
import asyncio
import requests
from telegram import Update
from telegram.ext import Application, CommandHandler, ContextTypes
import logging
import json
from datetime import datetime

# Enable detailed logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

# CoinGecko API endpoint
API_URL = "https://api.coingecko.com/api/v3/coins/markets"
PARAMS = {
    "vs_currency": "usd",
    "order": "market_cap_desc",
    "per_page": 10,
    "page": 1,
    "sparkline": False
}

# Bot token
TOKEN = "8014754930:AAG3FxXaw4AAiJfP5dPmoV5HoqZirR6Qvwc"

def get_crypto_data():
    """Fetch cryptocurrency data from CoinGecko API"""
    try:
        logger.info("🔄 Fetching crypto data from CoinGecko...")
        response = requests.get(API_URL, params=PARAMS, timeout=15)
        
        logger.info(f"📡 API Response: {response.status_code}")
        
        if response.status_code == 200:
            data = response.json()
            # Filter out USDT and get top 5
            filtered_data = [crypto for crypto in data if crypto.get('symbol', '').upper() != 'USDT']
            result = filtered_data[:5]
            logger.info(f"✅ Successfully fetched {len(result)} cryptocurrencies")
            return result
        else:
            logger.error(f"❌ API Error: {response.status_code} - {response.text}")
            return []
            
    except Exception as e:
        logger.error(f"❌ Exception in get_crypto_data: {e}")
        return []

def format_crypto_message(cryptos):
    """Format cryptocurrency data into a message"""
    if not cryptos:
        return "❌ Unable to fetch cryptocurrency data. Please try again later."
    
    message = "🚀 **Top 5 Cryptocurrencies** 🚀\n\n"
    
    for i, crypto in enumerate(cryptos, 1):
        name = crypto.get('name', 'Unknown')
        symbol = crypto.get('symbol', '').upper()
        price = crypto.get('current_price', 0)
        change_24h = crypto.get('price_change_percentage_24h', 0)
        
        # Format price
        if price >= 1:
            price_str = f"${price:,.2f}"
        else:
            price_str = f"${price:.6f}"
        
        # Format change with emoji
        if change_24h > 0:
            change_str = f"📈 +{change_24h:.2f}%"
        elif change_24h < 0:
            change_str = f"📉 {change_24h:.2f}%"
        else:
            change_str = f"➡️ {change_24h:.2f}%"
        
        message += f"{i}. **{name}** ({symbol})\n"
        message += f"   💰 {price_str}\n"
        message += f"   📊 {change_str}\n\n"
    
    message += f"🔄 Updated at {datetime.now().strftime('%H:%M:%S')}"
    return message

async def send_crypto_update(context: ContextTypes.DEFAULT_TYPE):
    """Send crypto update to chat"""
    try:
        logger.info("📊 Sending crypto update...")
        
        cryptos = get_crypto_data()
        message = format_crypto_message(cryptos)
        
        await context.bot.send_message(
            chat_id=context.job.chat_id,
            text=message,
            parse_mode='Markdown'
        )
        
        logger.info("✅ Crypto update sent successfully!")
        
    except Exception as e:
        logger.error(f"❌ Error sending crypto update: {e}")
        try:
            await context.bot.send_message(
                chat_id=context.job.chat_id,
                text="❌ Error updating crypto prices. Will retry soon."
            )
        except:
            pass

async def start_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handle /start command"""
    try:
        chat_id = update.effective_chat.id
        user_name = update.effective_user.first_name or "User"
        
        logger.info(f"🚀 /start command from {user_name} (chat_id: {chat_id})")
        
        # Remove existing jobs for this chat
        current_jobs = context.job_queue.get_jobs_by_name(str(chat_id))
        for job in current_jobs:
            job.schedule_removal()
            logger.info("🗑️ Removed existing job")
        
        # Schedule new job
        context.job_queue.run_repeating(
            send_crypto_update,
            interval=60,  # Every 60 seconds
            first=3,      # First update in 3 seconds
            chat_id=chat_id,
            name=str(chat_id)
        )
        
        welcome_msg = f"👋 Hello {user_name}!\n\n"
        welcome_msg += "🤖 **Crypto Price Bot** is now active!\n\n"
        welcome_msg += "📊 You'll receive top 5 crypto prices every minute\n"
        welcome_msg += "🔄 First update in 3 seconds...\n\n"
        welcome_msg += "**Commands:**\n"
        welcome_msg += "• `/start` - Start updates\n"
        welcome_msg += "• `/stop` - Stop updates\n"
        welcome_msg += "• `/prices` - Get prices now"
        
        await update.message.reply_text(welcome_msg, parse_mode='Markdown')
        logger.info("✅ Welcome message sent")
        
    except Exception as e:
        logger.error(f"❌ Error in start_command: {e}")
        await update.message.reply_text("❌ Error starting bot. Please try again.")

async def stop_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handle /stop command"""
    try:
        chat_id = update.effective_chat.id
        logger.info(f"🛑 /stop command from chat_id: {chat_id}")
        
        current_jobs = context.job_queue.get_jobs_by_name(str(chat_id))
        if current_jobs:
            for job in current_jobs:
                job.schedule_removal()
            await update.message.reply_text("🛑 **Updates stopped!** Use /start to resume.")
            logger.info("✅ Jobs stopped successfully")
        else:
            await update.message.reply_text("ℹ️ No active updates to stop.")
            
    except Exception as e:
        logger.error(f"❌ Error in stop_command: {e}")
        await update.message.reply_text("❌ Error stopping updates.")

async def prices_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handle /prices command - immediate update"""
    try:
        chat_id = update.effective_chat.id
        logger.info(f"💰 /prices command from chat_id: {chat_id}")
        
        await update.message.reply_text("🔄 Fetching current prices...")
        
        cryptos = get_crypto_data()
        message = format_crypto_message(cryptos)
        
        await update.message.reply_text(message, parse_mode='Markdown')
        logger.info("✅ Immediate prices sent")
        
    except Exception as e:
        logger.error(f"❌ Error in prices_command: {e}")
        await update.message.reply_text("❌ Error fetching prices.")

async def error_handler(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handle errors"""
    logger.error(f"❌ Update {update} caused error: {context.error}")

def main():
    """Main function"""
    try:
        logger.info("🚀 Starting Telegram Crypto Bot...")
        
        # Fix for Python 3.14 - ensure event loop exists
        try:
            loop = asyncio.get_event_loop()
            if loop.is_closed():
                raise RuntimeError("Event loop is closed")
        except RuntimeError:
            # Create new event loop if none exists or current one is closed
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            logger.info("✅ Created new event loop")
        
        # Test API first
        logger.info("🧪 Testing CoinGecko API...")
        test_data = get_crypto_data()
        if test_data:
            logger.info(f"✅ API test successful - got {len(test_data)} cryptos")
        else:
            logger.warning("⚠️ API test failed, but continuing...")
        
        # Create application
        application = Application.builder().token(TOKEN).build()
        
        # Add handlers
        application.add_handler(CommandHandler("start", start_command))
        application.add_handler(CommandHandler("stop", stop_command))
        application.add_handler(CommandHandler("prices", prices_command))
        application.add_error_handler(error_handler)
        
        logger.info("✅ Bot handlers registered")
        logger.info("🔄 Starting polling...")
        
        # Start the bot
        application.run_polling(
            allowed_updates=Update.ALL_TYPES,
            drop_pending_updates=True
        )
        
    except Exception as e:
        logger.error(f"❌ Failed to start bot: {e}")
        raise

if __name__ == "__main__":
    main()